contracts/governance/GoodDollarStaking.sol

```diff
@@ -0,0 +1,370 @@
+
+	/**
+	 * @dev Emitted when `staker` withdraws an `amount` of staked GoodDollars
+	 */
+	event StakeWithdraw(
+		address indexed staker,
+		uint256 amount, //amount withdrawn including gdRewards
+		uint256 goodRewards,
+		uint256 gdRewards
+	);
+
+	/**
+	 * @dev Constructor
+	 * @param _ns The address of the INameService contract
+	 */
+	constructor(
+		INameService _ns,
+		uint128 _interestRatePerBlock,
+		uint128 _numberOfBlocksPerYear,
+		uint32 _daysUntilUpgrade
+	) {
+		_setAPY(_interestRatePerBlock);
+		setDAO(_ns);
+		numberOfBlocksPerYear = _numberOfBlocksPerYear;
+		token = ERC20(nameService.getAddress("GOODDOLLAR"));
+		__ERC20_init("G$ Savings", "svG$");
+		rewardsPerBlock[address(this)] = 0;
+		createdAt = uint128(block.timestamp);
+		daysUntilUpgrade = _daysUntilUpgrade;
+	}
+
+	function getChainBlocksPerMonth() public view override returns (uint256) {
+		return numberOfBlocksPerYear / 12;
+	}
+
+	/**
+	 * @dev Allows a staker to deposit Tokens. Notice that `approve` is
+	 * needed to be executed before the execution of this method.
+	 * Can be executed only when the contract is not paused.
+	 * @param _amount The amount of GD to stake
+	 * @param _donationRatio percentage between 0-100
+	 */
+	function stake(uint256 _amount, uint32 _donationRatio) external {
+		require(
+			token.transferFrom(_msgSender(), address(this), _amount),
+			"transferFrom failed, make sure you approved token transfer"
+		);
+		_stakeFrom(_msgSender(), _amount, _donationRatio);
+	}
+
+	function _stakeFrom(
+		address _from,
+		uint256 _amount,
+		uint32 _donationRatio
+	) internal {
+		/* GOOD rewards Updates */
+		_increaseProductivity(address(this), _from, _amount, 0, block.number);
+		_mint(_from, _amount); // mint Staking token for staker
+		_mintGOODRewards(_from);
+		/* end GOOD rewards Updates */
+
+		/* G$ rewards updates */
+		_stake(_from, _amount, _donationRatio); //this will validate _amount and _donation ratio
+
+		emit Staked(_from, _amount, _donationRatio);
+	}
+
+	function onTokenTransfer(
+		address _from,
+		uint256 _amount,
+		bytes calldata data
+	) external returns (bool success) {
+		require(_msgSender() == address(token), "unsupported token");
+		uint32 donationRatio = abi.decode(data, (uint32));
+
+		_stakeFrom(_from, _amount, donationRatio);
+		return true;
+	}
+
+	/**
+	 * @dev Withdraws _amount from the staker principle
+	 * we use getPrinciple and not getProductivity because the user can have more G$ to withdraw than he staked
+	 */
+	function withdrawStake(uint256 _amount)
+		public
+		nonReentrant
+		returns (uint256 goodRewards, uint256 gdRewards)
+	{
+		uint256 depositComponent;
+
+		//in case amount is 0 this will just withdraw the GOOD rewards, this is required for withdrawRewards
+		if (_amount > 0) {
+			/* G$ rewards update */
+			//we get the relative part user is withdrawing from his original deposit, his principle is composed of deposit+earned interest
```

Comment:
Here and on the function comment it says principle, when we use withdraw (that behinds the scenes uses principle so maybe its ok)